File: mesop/components/icon/icon.py
import mesop.components.icon.icon_pb2 as icon_pb
from mesop.component_helpers import (
  Style,
  insert_component,
  register_native_component,
)


@register_native_component
def icon(
  icon: str | None = None,
  *,
  key: str | None = None,
  style: Style | None = None,
):
  """Creates a Icon component.

  Args:
    key: The component [key](../components/index.md#component-key).
    icon: Name of the [Material Symbols icon](https://fonts.google.com/icons).
    style: Inline styles
  """
  insert_component(
    key=key,
    type_name="icon",
    proto=icon_pb.IconType(
      icon=icon,
    ),
    style=style,
  )

File: mesop/components/icon/icon.ts
import {MatIconModule} from '@angular/material/icon';
import {Component, Input} from '@angular/core';
import {
  Key,
  Style,
  Type,
} from 'mesop/mesop/protos/ui_jspb_proto_pb/mesop/protos/ui_pb';
import {IconType} from 'mesop/mesop/components/icon/icon_jspb_proto_pb/mesop/components/icon/icon_pb';
import {Channel} from '../../web/src/services/channel';
import {formatStyle} from '../../web/src/utils/styles';

@Component({
  templateUrl: 'icon.ng.html',
  standalone: true,
  imports: [MatIconModule],
})
export class IconComponent {
  @Input({required: true}) type!: Type;
  @Input() key!: Key;
  @Input() style!: Style;
  private _config!: IconType;

  constructor(private readonly channel: Channel) {}

  ngOnChanges() {
    this._config = IconType.deserializeBinary(
      this.type.getValue() as unknown as Uint8Array,
    );
  }

  config(): IconType {
    return this._config;
  }

  getStyle(): string {
    return formatStyle(this.style);
  }
}
